# Project: cleanning - [candy data]
### this notebook is for educational purpose . in this notebook we will do some cleanning

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">wrangling</a></li>
  
</ul>

<a id='intro'></a>
## Introduction

### Dataset Description 

> i will go through candy data 2017 , this data has :
- Internal_ID
- Q1-Going Out?: Are you actually going trick or treating yourself? "has values `Yes or No`"
- Q2-Gender: It has four different options, `Femal ,Male ,Other ,I'd rather not say`
- Q3-Age: Numerical field
- Q4-Country: Text Field, but users have written their own version of the names. Example, for America, there are entries such as USA, us, US, America so we should consider that while cleanning
- Q5-State/Province: Text Field, but users have written their own version of the names. Same as the country data.
- Q6-Joy Or Despair: All kinds of chocolate bars are the questions with three distinct options to choose from (Joy, Meh, Despair).
- Q7-Joy Other: Text Field to write items not included above that give you JOY. Lots of missing values.
- Q8-Despair Other: Text Field ti write items not included above that give you DESPAIR. Lots of missing values.
- Q9-Other Comments: Text Field. Lots of missing values.
- Q10-Dress: Binary field. Missing values present
- Unnamed: 113
- Q11-Day: Binary Answer Field. Missing values present


# 1- import packages and load data

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt  




df = pd.read_excel("candyhierarchy2017.xlsx")

c:\Users\User\anaconda3\Lib\site-packages\openpyxl\worksheet\_read_only.py:85: UserWarning: Unknown extension is not supported and will be removed
  for idx, row in parser.parse():


In [2]:
df.head()

Internal ID Q1: GOING OUT? Q2: GENDER Q3: AGE Q4: COUNTRY  \
0     90258773            NaN        NaN     NaN         NaN   
1     90272821             No       Male      44        USA    
2     90272829            NaN       Male      49         USA   
3     90272840             No       Male      40          us   
4     90272841             No       Male      23         usa   

  Q5: STATE, PROVINCE, COUNTY, ETC Q6 | 100 Grand Bar  \
0                              NaN                NaN   
1                               NM                MEH   
2                         Virginia                NaN   
3                               or                MEH   
4                         exton pa                JOY   

  Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                                NaN                                       
1                                            DESPAIR                                       
2                                                NaN                                       
3                                            DESPAIR                                       
4                                            DESPAIR                                       

  Q6 | Any full-sized candy bar Q6 | Black Jacks  ... Q8: DESPAIR OTHER  \
0                           NaN              NaN  ...               NaN   
1                           JOY              MEH  ...               NaN   
2                           NaN              NaN  ...               NaN   
3                           JOY              MEH  ...               NaN   
4                           JOY          DESPAIR  ...               NaN   

                                  Q9: OTHER COMMENTS      Q10: DRESS  \
0                                                NaN             NaN   
1  Bottom line is Twix is really the only candy w...  White and gold   
2                                                NaN             NaN   
3                             Raisins can go to hell  White and gold   
4                                                NaN  White and gold   

  Unnamed: 113 Q11: DAY Q12: MEDIA [Daily Dish] Q12: MEDIA [Science]  \
0          NaN      NaN                     NaN                  NaN   
1          NaN   Sunday                     NaN                  1.0   
2          NaN      NaN                     NaN                  NaN   
3          NaN   Sunday                     NaN                  1.0   
4          NaN   Friday                     NaN                  1.0   

  Q12: MEDIA [ESPN] Q12: MEDIA [Yahoo] Click Coordinates (x, y)  
0               NaN                NaN                      NaN  
1               NaN                NaN                 (84, 25)  
2               NaN                NaN                      NaN  
3               NaN                NaN                 (75, 23)  
4               NaN                NaN                 (70, 10)  

[5 rows x 120 columns]

<a id='wrangling'></a>
## Data Wrangling



### General Properties

# 2- show some general properties for the data 

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2460 entries, 0 to 2459
Columns: 120 entries, Internal ID to Click Coordinates (x, y)
dtypes: float64(4), int64(1), object(115)
memory usage: 2.3+ MB


In [4]:
df.describe()

Internal ID  Q12: MEDIA [Daily Dish]  Q12: MEDIA [Science]  \
count  2.460000e+03                     85.0                1362.0   
mean   9.028016e+07                      1.0                   1.0   
std    7.435532e+03                      0.0                   0.0   
min    9.025877e+07                      1.0                   1.0   
25%    9.027520e+07                      1.0                   1.0   
50%    9.027780e+07                      1.0                   1.0   
75%    9.028242e+07                      1.0                   1.0   
max    9.031480e+07                      1.0                   1.0   

       Q12: MEDIA [ESPN]  Q12: MEDIA [Yahoo]  
count               99.0                67.0  
mean                 1.0                 1.0  
std                  0.0                 0.0  
min                  1.0                 1.0  
25%                  1.0                 1.0  
50%                  1.0                 1.0  
75%                  1.0                 1.0  
max                  1.0                 1.0

# 3- write summary about what you get form the above functions
> #### tips : this summary can be about : if there is any missing values , if there a wrong data type , if there a wrong range of values , ....

RangeIndex: 2460 entries, 0 to 2459

Columns: 120 entries, Internal ID to Click Coordinates (x, y)

dtypes: float64(4), int64(1), object(115)

# cleaning

### 4- drop all the useless columns and  rows with too much NaN , 
> tip : useless columns can be like `Internal ID`and can be also columns with too much NaN 
- rename columns for easy access

In [5]:
df.drop(['Internal ID' ,'Q7: JOY OTHER', 'Q8: DESPAIR OTHER' ,'Unnamed: 113' , 'Q12: MEDIA [Daily Dish]' 
         , 'Q12: MEDIA [ESPN]' , 'Q12: MEDIA [Yahoo]' , 'Q12: MEDIA [Science]' 
         ,'Click Coordinates (x, y)' , 'Q9: OTHER COMMENTS'  ] , axis=1 , inplace=True)

In [6]:

for col in df.columns:
    new_col_name = col.replace('Q6 | ','')
    df.rename(columns={col : new_col_name}, inplace=True)

In [7]:
df.rename(columns={'Q1: GOING OUT?':'GOING OUT','Q2: GENDER':'GENDER','Q3: AGE':'AGE','Q4: COUNTRY':'COUNTRY',
                   'Q5: STATE, PROVINCE, COUNTY, ETC':'AREA','Q10: DRESS':'DRESS',
                   'Q11: DAY':'DAY'},inplace=True)

In [8]:
df.head(5)

GOING OUT GENDER  AGE COUNTRY      AREA 100 Grand Bar  \
0       NaN    NaN  NaN     NaN       NaN           NaN   
1        No   Male   44    USA         NM           MEH   
2       NaN   Male   49     USA  Virginia           NaN   
3        No   Male   40      us        or           MEH   
4        No   Male   23     usa  exton pa           JOY   

  Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                                NaN                                  
1                                            DESPAIR                                  
2                                                NaN                                  
3                                            DESPAIR                                  
4                                            DESPAIR                                  

  Any full-sized candy bar Black Jacks Bonkers (the candy)  ... Trail Mix  \
0                      NaN         NaN                 NaN  ...       NaN   
1                      JOY         MEH             DESPAIR  ...   DESPAIR   
2                      NaN         NaN                 NaN  ...       NaN   
3                      JOY         MEH                 MEH  ...       MEH   
4                      JOY     DESPAIR                 MEH  ...   DESPAIR   

  Twix Vials of pure high fructose corn syrup, for main-lining into your vein  \
0  NaN                                                NaN                       
1  JOY                                            DESPAIR                       
2  NaN                                                NaN                       
3  JOY                                            DESPAIR                       
4  JOY                                                MEH                       

   Vicodin Whatchamacallit Bars White Bread Whole Wheat anything  \
0      NaN                  NaN         NaN                  NaN   
1  DESPAIR              DESPAIR     DESPAIR              DESPAIR   
2      NaN                  NaN         NaN                  NaN   
3      JOY                  JOY     DESPAIR              DESPAIR   
4      JOY                  JOY     DESPAIR              DESPAIR   

  York Peppermint Patties           DRESS     DAY  
0                     NaN             NaN     NaN  
1                 DESPAIR  White and gold  Sunday  
2                     NaN             NaN     NaN  
3                 DESPAIR  White and gold  Sunday  
4                     JOY  White and gold  Friday  

[5 rows x 110 columns]

### 5 - clean `going out` column 
> tip : you should choose which best value to put in place of these NaNs ( mode, unknown,other ....etc)

In [9]:
df['GOING OUT'].unique()

array([nan, 'No', 'Yes'], dtype=object)

In [10]:
df['GOING OUT'] = df['GOING OUT'].fillna('NOT SPECIFIED')

In [11]:
df.head(5)

GOING OUT GENDER  AGE COUNTRY      AREA 100 Grand Bar  \
0  NOT SPECIFIED    NaN  NaN     NaN       NaN           NaN   
1             No   Male   44    USA         NM           MEH   
2  NOT SPECIFIED   Male   49     USA  Virginia           NaN   
3             No   Male   40      us        or           MEH   
4             No   Male   23     usa  exton pa           JOY   

  Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                                NaN                                  
1                                            DESPAIR                                  
2                                                NaN                                  
3                                            DESPAIR                                  
4                                            DESPAIR                                  

  Any full-sized candy bar Black Jacks Bonkers (the candy)  ... Trail Mix  \
0                      NaN         NaN                 NaN  ...       NaN   
1                      JOY         MEH             DESPAIR  ...   DESPAIR   
2                      NaN         NaN                 NaN  ...       NaN   
3                      JOY         MEH                 MEH  ...       MEH   
4                      JOY     DESPAIR                 MEH  ...   DESPAIR   

  Twix Vials of pure high fructose corn syrup, for main-lining into your vein  \
0  NaN                                                NaN                       
1  JOY                                            DESPAIR                       
2  NaN                                                NaN                       
3  JOY                                            DESPAIR                       
4  JOY                                                MEH                       

   Vicodin Whatchamacallit Bars White Bread Whole Wheat anything  \
0      NaN                  NaN         NaN                  NaN   
1  DESPAIR              DESPAIR     DESPAIR              DESPAIR   
2      NaN                  NaN         NaN                  NaN   
3      JOY                  JOY     DESPAIR              DESPAIR   
4      JOY                  JOY     DESPAIR              DESPAIR   

  York Peppermint Patties           DRESS     DAY  
0                     NaN             NaN     NaN  
1                 DESPAIR  White and gold  Sunday  
2                     NaN             NaN     NaN  
3                 DESPAIR  White and gold  Sunday  
4                     JOY  White and gold  Friday  

[5 rows x 110 columns]

### 6 - clean ` gender` column 
> tip : you should choose which best value to put in place of these NaNs ( mode, unknown,other ....etc)

In [12]:
df['GENDER'].unique()

array([nan, 'Male', 'Female', "I'd rather not say", 'Other'], dtype=object)

In [13]:
df['GENDER'] = df['GENDER'].fillna("I'd rather not say")

In [14]:
df['GENDER'].replace("Other", "I'd rather not say", inplace=True)

In [15]:
df.head(5)

GOING OUT              GENDER  AGE COUNTRY      AREA 100 Grand Bar  \
0  NOT SPECIFIED  I'd rather not say  NaN     NaN       NaN           NaN   
1             No                Male   44    USA         NM           MEH   
2  NOT SPECIFIED                Male   49     USA  Virginia           NaN   
3             No                Male   40      us        or           MEH   
4             No                Male   23     usa  exton pa           JOY   

  Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                                NaN                                  
1                                            DESPAIR                                  
2                                                NaN                                  
3                                            DESPAIR                                  
4                                            DESPAIR                                  

  Any full-sized candy bar Black Jacks Bonkers (the candy)  ... Trail Mix  \
0                      NaN         NaN                 NaN  ...       NaN   
1                      JOY         MEH             DESPAIR  ...   DESPAIR   
2                      NaN         NaN                 NaN  ...       NaN   
3                      JOY         MEH                 MEH  ...       MEH   
4                      JOY     DESPAIR                 MEH  ...   DESPAIR   

  Twix Vials of pure high fructose corn syrup, for main-lining into your vein  \
0  NaN                                                NaN                       
1  JOY                                            DESPAIR                       
2  NaN                                                NaN                       
3  JOY                                            DESPAIR                       
4  JOY                                                MEH                       

   Vicodin Whatchamacallit Bars White Bread Whole Wheat anything  \
0      NaN                  NaN         NaN                  NaN   
1  DESPAIR              DESPAIR     DESPAIR              DESPAIR   
2      NaN                  NaN         NaN                  NaN   
3      JOY                  JOY     DESPAIR              DESPAIR   
4      JOY                  JOY     DESPAIR              DESPAIR   

  York Peppermint Patties           DRESS     DAY  
0                     NaN             NaN     NaN  
1                 DESPAIR  White and gold  Sunday  
2                     NaN             NaN     NaN  
3                 DESPAIR  White and gold  Sunday  
4                     JOY  White and gold  Friday  

[5 rows x 110 columns]

### 7 - clean ` age` column 
> tip : note the type of age column , and NaN values
 - try to use `to_numeric` 
 - tip : you should choose which best value to put in place of these NaNs ( mean, median, ....etc)

In [16]:
df['AGE'].isna().sum()

84

In [17]:
df['AGE'] = pd.to_numeric(df['AGE'], errors='coerce')


In [18]:
df['AGE'].fillna(df['AGE'].mean(), inplace=True)

In [19]:
df['AGE'] = df['AGE'].astype(int)

In [20]:
df.loc[df['AGE'] > 100 ] = np.nan

In [21]:
df.head()

GOING OUT              GENDER   AGE COUNTRY      AREA 100 Grand Bar  \
0  NOT SPECIFIED  I'd rather not say  42.0     NaN       NaN           NaN   
1             No                Male  44.0    USA         NM           MEH   
2  NOT SPECIFIED                Male  49.0     USA  Virginia           NaN   
3             No                Male  40.0      us        or           MEH   
4             No                Male  23.0     usa  exton pa           JOY   

  Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                                NaN                                  
1                                            DESPAIR                                  
2                                                NaN                                  
3                                            DESPAIR                                  
4                                            DESPAIR                                  

  Any full-sized candy bar Black Jacks Bonkers (the candy)  ... Trail Mix  \
0                      NaN         NaN                 NaN  ...       NaN   
1                      JOY         MEH             DESPAIR  ...   DESPAIR   
2                      NaN         NaN                 NaN  ...       NaN   
3                      JOY         MEH                 MEH  ...       MEH   
4                      JOY     DESPAIR                 MEH  ...   DESPAIR   

  Twix Vials of pure high fructose corn syrup, for main-lining into your vein  \
0  NaN                                                NaN                       
1  JOY                                            DESPAIR                       
2  NaN                                                NaN                       
3  JOY                                            DESPAIR                       
4  JOY                                                MEH                       

   Vicodin Whatchamacallit Bars White Bread Whole Wheat anything  \
0      NaN                  NaN         NaN                  NaN   
1  DESPAIR              DESPAIR     DESPAIR              DESPAIR   
2      NaN                  NaN         NaN                  NaN   
3      JOY                  JOY     DESPAIR              DESPAIR   
4      JOY                  JOY     DESPAIR              DESPAIR   

  York Peppermint Patties           DRESS     DAY  
0                     NaN             NaN     NaN  
1                 DESPAIR  White and gold  Sunday  
2                     NaN             NaN     NaN  
3                 DESPAIR  White and gold  Sunday  
4                     JOY  White and gold  Friday  

[5 rows x 110 columns]

### 8 - clean ` country` column
> tip : note that country column has country names with different writing ways 
  - use fuzzywuzzy to deal with this problem
  - you should choose which best value to put in place of these NaNs ( mode, unknown,other ....etc)

In [22]:
df['COUNTRY'].unique()

 

array([nan, 'USA ', 'USA', 'us', 'usa', 'canada', 'Canada', 'Us', 'US',
       'Murica', 'United States', 'uk', 'United Kingdom', 'united states',
       'Usa', 'United States ', 'United staes',
       'United States of America', 'UAE', 'England', 'UK', 'canada ',
       'Mexico', 'United states', 'u.s.a.', 'USAUSAUSA', 'america', 35,
       'france', 'United States of America ', 'U.S.A.', 'finland',
       'unhinged states', 'Canada ', 'united states of america',
       'US of A', 'Unites States', 'The United States', 'North Carolina ',
       'Unied States', 'Netherlands', 'germany', 'Europe', 'Earth', 'U S',
       'u.s.', 'U.K. ', 'Costa Rica', 'The United States of America',
       'unite states', 'U.S.', 46, 'cascadia', 'Australia',
       'insanity lately', 'Greece', 'USA? Hard to tell anymore..',
       "'merica", 'usas', 'Pittsburgh', 45, 'United State', 32, 'France',
       'australia', 'Can', 'Canae', 'New York', 'Trumpistan', 'Ireland',
       'United Sates', 'Korea', 'Cali

In [23]:
df['COUNTRY'].value_counts()

COUNTRY
USA                  699
United States        497
usa                  217
Canada               178
Usa                  139
                    ... 
Korea                  1
United Sates           1
unhinged states        1
Trumpistan             1
Fear and Loathing      1
Name: count, Length: 126, dtype: int64

In [24]:

df['COUNTRY'] = df['COUNTRY'].str.upper()
df['COUNTRY'] = df['COUNTRY'].str.strip()

In [25]:

mapping = {'US':'USA','UNITED STATES':'USA','UNITED STATES OF AMERICA':'USA','U.S.':'USA',
           'U.S.A.':'USA','AMERICA':'USA','UNITES STATES':'USA','US OF A':'USA','THE UNITED STATES':'USA',
           'UNITED STATED':'USA','UNITED STATE':'USA','UNITED STSTES':'USA','UNITED KINGDOM':'UK'}
df['COUNTRY'] = df['COUNTRY'].replace(mapping)

In [26]:
bad_values = df['COUNTRY'].value_counts()[df['COUNTRY'].value_counts().values < 2].index
final = df['COUNTRY'].isin(bad_values)

In [27]:

df['COUNTRY'].loc[final] = 'USA'

C:\Users\User\AppData\Local\Temp\ipykernel_8844\1404828291.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['COUNTRY'].loc[final] = 'USA'


In [28]:
df['COUNTRY'].fillna('USA',inplace=True)

In [29]:

df['COUNTRY'].value_counts()

COUNTRY
USA            2150
CANADA          223
UK               32
GERMANY          10
AUSTRALIA         7
NETHERLANDS       6
JAPAN             5
SCOTLAND          5
MEXICO            4
IRELAND           4
FRANCE            3
SWITZERLAND       3
DENMARK           2
CHINA             2
HONG KONG         2
SWEDEN            2
Name: count, dtype: int64

### 9 - clean ` area` column
> tip : area column has the same problem as country columns , but looking at all it's unique values may give us another way to deal with it rather than try to fix it

In [30]:
df['AREA'].unique()

array([nan, 'NM', 'Virginia', 'or', 'exton pa', 'Colorado ', 'ontario',
       'Ontario', 'Wa', 'Quebec', 'NY', 'California', 'Texas', 'NEW YORK',
       'CO', 'IN', 'MN', 'aberdeenshire', 'Alberta', 'Wisconsin, 53705',
       'RI', 'NC', 'Scotland', 'minnesota', 'Michigan', 'Florida',
       'New Jersey', 'VA', 'Colorado', 'Los Angeles, California', 'CA',
       'New York ', 'Chicago, Illinois', 'Nj', 'Wisconsin', 'Minnesota ',
       'London', 'mn', 'TEXAS', 'Illinois', 'District of Columbia',
       'Washington', 'Massachusetts ', 'OR', 'Montana', 'New York',
       'massachusetts', 'Abu Dhabi', 'TX', 'Arizona', 'Dc', 'yukon',
       'Pennsylvania ', 'MD', 'Ohio', 'Alberta ', 'OXFORDSHIRE', 'DC',
       'Iowa', 'Mn', 'TN', 'BC', 'Maryland', 'MI', 'Wisconsin ', 'alaska',
       'Ny', 'Colorado, Jefferson county ', 'Louisiana', 'pa',
       'CO, Larimer County', 'Nebraska ', 'Oregon', 'UT', 'ny',
       'Pennsylvania', 'CDMX', 'north carolina', 'Tennessee ', 'Texas ',
       'Knox cou

In [31]:
df.drop('AREA', axis=1, inplace=True)

In [32]:
df.head(5)

GOING OUT              GENDER   AGE COUNTRY 100 Grand Bar  \
0  NOT SPECIFIED  I'd rather not say  42.0     USA           NaN   
1             No                Male  44.0     USA           MEH   
2  NOT SPECIFIED                Male  49.0     USA           NaN   
3             No                Male  40.0     USA           MEH   
4             No                Male  23.0     USA           JOY   

  Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                                NaN                                  
1                                            DESPAIR                                  
2                                                NaN                                  
3                                            DESPAIR                                  
4                                            DESPAIR                                  

  Any full-sized candy bar Black Jacks Bonkers (the candy)  \
0                      NaN         NaN                 NaN   
1                      JOY         MEH             DESPAIR   
2                      NaN         NaN                 NaN   
3                      JOY         MEH                 MEH   
4                      JOY     DESPAIR                 MEH   

  Bonkers (the board game)  ... Trail Mix Twix  \
0                      NaN  ...       NaN  NaN   
1                  DESPAIR  ...   DESPAIR  JOY   
2                      NaN  ...       NaN  NaN   
3                  DESPAIR  ...       MEH  JOY   
4                  DESPAIR  ...   DESPAIR  JOY   

  Vials of pure high fructose corn syrup, for main-lining into your vein  \
0                                                NaN                       
1                                            DESPAIR                       
2                                                NaN                       
3                                            DESPAIR                       
4                                                MEH                       

   Vicodin Whatchamacallit Bars White Bread Whole Wheat anything  \
0      NaN                  NaN         NaN                  NaN   
1  DESPAIR              DESPAIR     DESPAIR              DESPAIR   
2      NaN                  NaN         NaN                  NaN   
3      JOY                  JOY     DESPAIR              DESPAIR   
4      JOY                  JOY     DESPAIR              DESPAIR   

  York Peppermint Patties           DRESS     DAY  
0                     NaN             NaN     NaN  
1                 DESPAIR  White and gold  Sunday  
2                     NaN             NaN     NaN  
3                 DESPAIR  White and gold  Sunday  
4                     JOY  White and gold  Friday  

[5 rows x 109 columns]

### 10 - clean ` Q6` columns
>- tip : Q6 has all kinds of chocolate bars and questions with three distinct options to choose from (Joy, Meh, Despair).
- so all the column that has Q6 is a candy name , so you should go through all these column and you should choose which best value to put in place of these NaNs ( mode, unknown,other ....etc)

In [33]:
df.iloc[:,4:106].isnull().sum()

100 Grand Bar                                                                        750
Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)    737
Any full-sized candy bar                                                             675
Black Jacks                                                                          961
Bonkers (the candy)                                                                  995
                                                                                    ... 
Vials of pure high fructose corn syrup, for main-lining into your vein               795
Vicodin                                                                              792
Whatchamacallit Bars                                                                 826
White Bread                                                                          760
Whole Wheat anything                                                                 750
Length: 102, dtype: i

In [34]:
df = df.drop(df[df.iloc[:,4:106].isnull().all(axis=1)].index)

In [35]:
df.iloc[:,4:106].isnull().sum()

100 Grand Bar                                                                         89
Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)     76
Any full-sized candy bar                                                              14
Black Jacks                                                                          300
Bonkers (the candy)                                                                  334
                                                                                    ... 
Vials of pure high fructose corn syrup, for main-lining into your vein               134
Vicodin                                                                              131
Whatchamacallit Bars                                                                 165
White Bread                                                                           99
Whole Wheat anything                                                                  89
Length: 102, dtype: i

In [36]:
values = (df.iloc[:,4:106].isnull().sum()/(df.shape[0]))*100
values = pd.DataFrame(values,columns=['percentage'])

In [37]:
values.head(5)

percentage
100 Grand Bar                                         4.947193
Anonymous brown globs that come in black and or...    4.224569
Any full-sized candy bar                              0.778210
Black Jacks                                          16.675931
Bonkers (the candy)                                  18.565870

In [38]:
values['percentage'].sort_values(ascending=False)[0:15]

JoyJoy (Mit Iodine!)                           20.455809
Maynards                                       20.344636
Reggie Jackson Bar                             19.788772
Bonkers (the board game)                       19.344080
Sweetums (a friend to diabetes)                19.121734
Bonkers (the candy)                            18.565870
Black Jacks                                    16.675931
Chick-o-Sticks (we don’t know what that is)    16.008894
Kinder Happy Hippo                             15.953307
Senior Mints                                   15.786548
Abstained from M&M'ing.                        15.786548
Take 5                                         14.396887
Smarties (Commonwealth)                        13.507504
Spotted Dick                                   12.451362
Goo Goo Clusters                               12.284603
Name: percentage, dtype: float64

In [39]:
missing = (df.iloc[:,4:106].isnull().sum()/(df.shape[0]))*100
columns_to_drop = missing[missing > 15].index

In [40]:
columns_to_drop

Index(['Black Jacks', 'Bonkers (the candy)', 'Bonkers (the board game)',
       'Chick-o-Sticks (we don’t know what that is)', 'JoyJoy (Mit Iodine!)',
       'Senior Mints', 'Kinder Happy Hippo', 'Maynards',
       'Abstained from M&M'ing.', 'Reggie Jackson Bar',
       'Sweetums (a friend to diabetes)'],
      dtype='object')

In [41]:
filtered_df = df.drop(columns= columns_to_drop)

In [42]:
filtered_df.iloc[:,4:106].isnull().sum()

100 Grand Bar                                                                         89
Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)     76
Any full-sized candy bar                                                              14
Bottle Caps                                                                          107
Box'o'Raisins                                                                         30
                                                                                    ... 
White Bread                                                                           99
Whole Wheat anything                                                                  89
York Peppermint Patties                                                               48
DRESS                                                                                 90
DAY                                                                                   70
Length: 94, dtype: in

In [43]:
filtered_df['100 Grand Bar'].value_counts(sort=True,ascending=False).index[0]

'JOY'

In [44]:
filtered_df.shape

(1799, 98)

In [45]:
for col in filtered_df.iloc[:,4:96]:
    filtered_df[col] = filtered_df[col].fillna(filtered_df[col].value_counts(sort=True,ascending=False).index[0])

In [46]:
filtered_df.iloc[:,4:96].isnull().sum().sum()

0

In [47]:
data = []
for col in filtered_df.iloc[:,4:96]:
    num_joy = df[col].value_counts()['JOY']
    num_meh = df[col].value_counts()['MEH']
    num_despaire = df[col].value_counts()['DESPAIR']
    data.append({'Candy': col, 'JOY': num_joy, 'MEH': num_meh, 'DESPAIR': num_despaire})
data = pd.DataFrame(data)
data = data.set_index(data['Candy'])

In [48]:
data.head()

Candy  \
Candy                                                                                                   
100 Grand Bar                                                                           100 Grand Bar   
Anonymous brown globs that come in black and or...  Anonymous brown globs that come in black and o...   
Any full-sized candy bar                                                     Any full-sized candy bar   
Bottle Caps                                                                               Bottle Caps   
Box'o'Raisins                                                                           Box'o'Raisins   

                                                     JOY  MEH  DESPAIR  
Candy                                                                   
100 Grand Bar                                        873  752       85  
Anonymous brown globs that come in black and or...   175  460     1088  
Any full-sized candy bar                            1556  212       17  
Bottle Caps                                          465  668      559  
Box'o'Raisins                                        117  475     1177

In [49]:
data['JOY'].sort_values(ascending=False)[0:10]

Candy
Any full-sized candy bar                1556
Reese’s Peanut Butter Cups              1496
Kit Kat                                 1445
Cash, or other forms of legal tender    1438
Twix                                    1411
Snickers                                1396
Tolberone something or other            1318
Peanut M&M’s                            1289
Lindt Truffle                           1275
Nestle Crunch                           1186
Name: JOY, dtype: int64

In [50]:
filtered_df['Butterfinger'].value_counts()['JOY']

1198

In [52]:
for col in filtered_df.iloc[:,4:96]:
    filtered_df[col].value_counts()

### 11 - clean ` dress` column
>tip : you should choose which best value to put in place of these NaNs ( mode, unknown,other ....etc)

In [53]:
df['DRESS'].unique()

array(['White and gold', nan, 'Blue and black'], dtype=object)

In [54]:
df['DRESS']=df['DRESS'].fillna('Prefer not to say')

In [55]:
df.head(5)

GOING OUT GENDER   AGE COUNTRY 100 Grand Bar  \
1        No   Male  44.0     USA           MEH   
3        No   Male  40.0     USA           MEH   
4        No   Male  23.0     USA           JOY   
5        No   Male  42.0     USA           JOY   
7        No   Male  33.0  CANADA           JOY   

  Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
1                                            DESPAIR                                  
3                                            DESPAIR                                  
4                                            DESPAIR                                  
5                                            DESPAIR                                  
7                                            DESPAIR                                  

  Any full-sized candy bar Black Jacks Bonkers (the candy)  \
1                      JOY         MEH             DESPAIR   
3                      JOY         MEH                 MEH   
4                      JOY     DESPAIR                 MEH   
5                      JOY         NaN                 NaN   
7                      JOY     DESPAIR             DESPAIR   

  Bonkers (the board game)  ... Trail Mix Twix  \
1                  DESPAIR  ...   DESPAIR  JOY   
3                  DESPAIR  ...       MEH  JOY   
4                  DESPAIR  ...   DESPAIR  JOY   
5                      NaN  ...       MEH  JOY   
7                      MEH  ...   DESPAIR  JOY   

  Vials of pure high fructose corn syrup, for main-lining into your vein  \
1                                            DESPAIR                       
3                                            DESPAIR                       
4                                                MEH                       
5                                            DESPAIR                       
7                                                JOY                       

   Vicodin Whatchamacallit Bars White Bread Whole Wheat anything  \
1  DESPAIR              DESPAIR     DESPAIR              DESPAIR   
3      JOY                  JOY     DESPAIR              DESPAIR   
4      JOY                  JOY     DESPAIR              DESPAIR   
5  DESPAIR                  JOY     DESPAIR              DESPAIR   
7      MEH              DESPAIR     DESPAIR              DESPAIR   

  York Peppermint Patties              DRESS     DAY  
1                 DESPAIR     White and gold  Sunday  
3                 DESPAIR     White and gold  Sunday  
4                     JOY     White and gold  Friday  
5                     JOY  Prefer not to say     NaN  
7                 DESPAIR     Blue and black  Friday  

[5 rows x 109 columns]

### 12 - clean ` day` column
>tip : you should choose which best value to put in place of these NaNs ( mode, unknown,other ....etc)

In [56]:
df['DAY'].unique()


array(['Sunday', 'Friday', nan], dtype=object)

In [57]:
df['DAY']=df['DAY'].fillna('Other')

In [58]:
df.head(5)

GOING OUT GENDER   AGE COUNTRY 100 Grand Bar  \
1        No   Male  44.0     USA           MEH   
3        No   Male  40.0     USA           MEH   
4        No   Male  23.0     USA           JOY   
5        No   Male  42.0     USA           JOY   
7        No   Male  33.0  CANADA           JOY   

  Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
1                                            DESPAIR                                  
3                                            DESPAIR                                  
4                                            DESPAIR                                  
5                                            DESPAIR                                  
7                                            DESPAIR                                  

  Any full-sized candy bar Black Jacks Bonkers (the candy)  \
1                      JOY         MEH             DESPAIR   
3                      JOY         MEH                 MEH   
4                      JOY     DESPAIR                 MEH   
5                      JOY         NaN                 NaN   
7                      JOY     DESPAIR             DESPAIR   

  Bonkers (the board game)  ... Trail Mix Twix  \
1                  DESPAIR  ...   DESPAIR  JOY   
3                  DESPAIR  ...       MEH  JOY   
4                  DESPAIR  ...   DESPAIR  JOY   
5                      NaN  ...       MEH  JOY   
7                      MEH  ...   DESPAIR  JOY   

  Vials of pure high fructose corn syrup, for main-lining into your vein  \
1                                            DESPAIR                       
3                                            DESPAIR                       
4                                                MEH                       
5                                            DESPAIR                       
7                                                JOY                       

   Vicodin Whatchamacallit Bars White Bread Whole Wheat anything  \
1  DESPAIR              DESPAIR     DESPAIR              DESPAIR   
3      JOY                  JOY     DESPAIR              DESPAIR   
4      JOY                  JOY     DESPAIR              DESPAIR   
5  DESPAIR                  JOY     DESPAIR              DESPAIR   
7      MEH              DESPAIR     DESPAIR              DESPAIR   

  York Peppermint Patties              DRESS     DAY  
1                 DESPAIR     White and gold  Sunday  
3                 DESPAIR     White and gold  Sunday  
4                     JOY     White and gold  Friday  
5                     JOY  Prefer not to say   Other  
7                 DESPAIR     Blue and black  Friday  

[5 rows x 109 columns]

### 13 - check
> after finish cleaning you should make sure your data is clean and that :
- there is no wrong data type
- there is no NaN values 
- all the column are cleaing and ready for the next step `EDA`

In [59]:
df.isna().sum()

GOING OUT                   0
GENDER                      0
AGE                         0
COUNTRY                     0
100 Grand Bar              89
                           ..
White Bread                99
Whole Wheat anything       89
York Peppermint Patties    48
DRESS                       0
DAY                         0
Length: 109, dtype: int64